File: api/controllers/profileController.js
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();
const ImageKit = require('imagekit');
const multer = require('multer');
const bcrypt = require('bcrypt');

// Configure ImageKit
const imagekit = new ImageKit({
  publicKey: process.env.IMAGEKIT_PUBLIC_KEY,
  privateKey: process.env.IMAGEKIT_PRIVATE_KEY,
  urlEndpoint: process.env.IMAGEKIT_URL_ENDPOINT
});

const updateProfilePhoto = async (req, res) => {
  try {
    const userId = req.user.id;
    
    if (!req.file) {
      return res.status(400).json({ message: 'No file uploaded' });
    }

    // Upload file to ImageKit
    const result = await imagekit.upload({
      file: req.file.buffer,
      fileName: `profile-${userId}-${Date.now()}`,
      folder: '/whoosh/profile_photos',
    });

    // Update user's photoUrl in database
    const updatedUser = await prisma.user.update({
      where: { id: userId },
      data: { photoUrl: result.url },
      select: {
        id: true,
        name: true,
        email: true,
        phoneNumber: true,
        photoUrl: true,
        role: true,
      },
    });

    res.json({
      message: 'Profile photo updated successfully',
      user: updatedUser,
    });
  } catch (error) {
    console.error('Profile photo update error:', error);
    res.status(500).json({ message: 'Failed to update profile photo' });
  }
};

const getProfile = async (req, res) => {
  try {
    const userId = req.user.id;

    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        name: true,
        email: true,
        phoneNumber: true,
        photoUrl: true,
        role: true,
      },
    });

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    res.json({ user });
  } catch (error) {
    console.error('Get profile error:', error);
    res.status(500).json({ message: 'Failed to fetch profile' });
  }
};

const updatePassword = async (req, res) => {
  try {
    const userId = req.user.id;
    const { currentPassword, newPassword, confirmPassword } = req.body;

    // Validate request body
    if (!currentPassword || !newPassword || !confirmPassword) {
      return res.status(400).json({ message: 'All fields are required' });
    }

    // Check if passwords match
    if (newPassword !== confirmPassword) {
      return res.status(400).json({ message: 'New passwords do not match' });
    }

    // Validate password strength
    if (newPassword.length < 8) {
      return res.status(400).json({ message: 'Password must be at least 8 characters long' });
    }

    // Get current user with password
    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        password: true,
      },
    });

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Verify current password
    const isPasswordValid = await bcrypt.compare(currentPassword, user.password);
    if (!isPasswordValid) {
      return res.status(401).json({ message: 'Current password is incorrect' });
    }

    // Hash new password
    const hashedPassword = await bcrypt.hash(newPassword, 10);

    // Update password in database
    await prisma.user.update({
      where: { id: userId },
      data: { password: hashedPassword },
    });

    res.json({ message: 'Password updated successfully' });
  } catch (error) {
    console.error('Password update error:', error);
    res.status(500).json({ message: 'Failed to update password' });
  }
};

module.exports = {
  updateProfilePhoto,
  getProfile,
  updatePassword,
};

File: lib/controllers/profile_controller.dart
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:get/get.dart';
import 'package:image_picker/image_picker.dart';
import 'package:get_storage/get_storage.dart';
import 'package:woosh/services/api_service.dart';
import 'package:woosh/models/user_model.dart';

class ProfileController extends GetxController {
  final storage = GetStorage();

  Rx<XFile?> selectedImage = Rx<XFile?>(null);
  RxBool isLoading = false.obs;
  RxString userName = ''.obs;
  RxString userEmail = ''.obs;
  RxString userPhone = ''.obs;
  RxString photoUrl = ''.obs;
  RxString userRole = ''.obs;
  RxString userRegion = ''.obs;
  RxString userDepartment = ''.obs;

  // Password update fields
  final RxBool isPasswordUpdating = false.obs;
  final RxString passwordError = ''.obs;
  final RxString passwordSuccess = ''.obs;

  @override
  void onInit() {
    super.onInit();
    loadUserData();
    fetchProfile();
  }

  void loadUserData() {
    final userData = storage.read('salesRep');
    if (userData != null) {
      userName.value = userData['name'] ?? '';
      userEmail.value = userData['email'] ?? '';
      userPhone.value = userData['phoneNumber'] ?? '';
      photoUrl.value = userData['photoUrl'] ?? '';
      userRole.value = userData['role'] ?? '';
      userRegion.value = userData['region'] ?? '';
      userDepartment.value = userData['department'] ?? '';
    }
  }

  Future<void> fetchProfile() async {
    try {
      print('🔄 Fetching profile data...');
      final response = await ApiService.getProfile();
      final userData = response['salesRep'];

      if (userData != null) {
        userName.value = userData['name'] ?? '';
        userEmail.value = userData['email'] ?? '';
        userPhone.value = userData['phoneNumber'] ?? '';
        photoUrl.value = userData['photoUrl'] ?? '';
        userRole.value = userData['role'] ?? '';
        userRegion.value = userData['region'] ?? '';
        userDepartment.value = userData['department'] ?? '';

        // Update storage with full user data
        storage.write('salesRep', userData);

        print('✅ Profile data loaded successfully: ${userData.toString()}');
      } else {
        print('❌ No user data in response - using cached data');
        // Don't throw exception, just use cached data
        loadUserData();
      }
    } catch (e) {
      print('❌ Error fetching profile: $e');

      // Only show snackbar for specific errors, not network issues
      if (e.toString().contains('Session expired') ||
          e.toString().contains('Authentication failed')) {
        Get.snackbar(
          'Authentication Error',
          'Please log in again',
          snackPosition: SnackPosition.TOP,
          backgroundColor: Get.theme.colorScheme.error.withOpacity(0.1),
          colorText: Get.theme.colorScheme.error,
        );
      }

      // Always fall back to cached data
      print('🔄 Falling back to cached profile data...');
      loadUserData();
    }
  }

  Future<void> pickImage() async {
    try {
      final ImagePicker picker = ImagePicker();
      final XFile? image = await picker.pickImage(
        source: ImageSource.gallery,
        imageQuality: 80,
      );

      if (image != null) {
        selectedImage.value = image;
        await updateProfilePhoto();
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to pick image: ${e.toString()}',
        snackPosition: SnackPosition.TOP,
      );
    }
  }

  Future<void> updateProfilePhoto() async {
    if (selectedImage.value == null) return;

    try {
      isLoading.value = true;

      // Pass XFile directly to handle both web and mobile platforms
      final photoUrl =
          await ApiService.updateProfilePhoto(selectedImage.value!);
      this.photoUrl.value = photoUrl;

      // Update storage
      final storedUser =
          storage.read('salesRep') as Map<String, dynamic>? ?? {};
      storedUser['photoUrl'] = photoUrl;
      storage.write('salesRep', storedUser);

      Get.snackbar(
        'Success',
        'Profile photo updated successfully',
        snackPosition: SnackPosition.TOP,
      );
    } catch (e) {
      Get.snackbar(
        'Error',
        e.toString().replaceAll('Exception: ', ''),
        snackPosition: SnackPosition.TOP,
        backgroundColor: Get.theme.colorScheme.error.withOpacity(0.1),
        colorText: Get.theme.colorScheme.error,
      );
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> updatePassword({
    required String currentPassword,
    required String newPassword,
    required String confirmPassword,
  }) async {
    try {
      // Reset status messages
      passwordError.value = '';
      passwordSuccess.value = '';
      isPasswordUpdating.value = true;

      print('PROFILE CONTROLLER: Starting password update process');

      // Validate passwords
      if (currentPassword.isEmpty ||
          newPassword.isEmpty ||
          confirmPassword.isEmpty) {
        passwordError.value = 'All fields are required';
        print('PROFILE CONTROLLER: Validation failed - Empty fields');
        return;
      }

      if (newPassword != confirmPassword) {
        passwordError.value = 'New passwords do not match';
        print('PROFILE CONTROLLER: Validation failed - Passwords do not match');
        return;
      }

      if (newPassword.length < 8) {
        passwordError.value = 'Password must be at least 8 characters long';
        print('PROFILE CONTROLLER: Validation failed - Password too short');
        return;
      }

      print('PROFILE CONTROLLER: Validation passed, calling API service');

      // Call API to update password
      final result = await ApiService.updatePassword(
        currentPassword: currentPassword,
        newPassword: newPassword,
        confirmPassword: confirmPassword,
      );

      print('PROFILE CONTROLLER: API response received: $result');

      if (result['success']) {
        passwordSuccess.value = result['message'];
        print('PROFILE CONTROLLER: Password update successful');
      } else {
        passwordError.value = result['message'];
        print(
            'PROFILE CONTROLLER: Password update failed: ${result['message']}');
      }
    } catch (e) {
      print('PROFILE CONTROLLER: Exception during password update: $e');
      passwordError.value = 'An error occurred: ${e.toString()}';
    } finally {
      isPasswordUpdating.value = false;
      print('PROFILE CONTROLLER: Password update process completed');
    }
  }
}
